File: workspace/algo-class-assignments/java/main/by/grodnosoft/algoclass/week2/CountQuickSortComparisons.java
package by.grodnosoft.algoclass.week2;

import java.io.BufferedReader;
import java.io.DataInputStream;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;

/**
 * The file contains all of the integers between 1 and 10,000 (inclusive) in unsorted order (with no integer repeated). 
 * The integer in the i-th row of the file gives you the i-th entry of an input array.
 * Your task is to compute the total number of comparisons used to sort the given input file by QuickSort. 
 * As you know, the number of comparisons depends on which elements are chosen as pivots, so we'll ask you to explore three different pivoting rules.
 * You should not count comparisons one-by-one. Rather, when there is a recursive call on a subarray of length m, 
 * you should simply add m−1 to your running total of comparisons. (This is because the pivot element will be compared to each of the other m−1 elements in the subarray in this recursive call.)
 * 
 * WARNING: The Partition subroutine can be implemented in several different ways, and different implementations can give you differing numbers of comparisons. 
 * For this problem, you should implement the Partition subroutine as it is described in the video lectures (otherwise you might get the wrong answer).
 * 
 * DIRECTIONS FOR PROBLEMS:
 * - For the first part of the programming assignment, you should always use the first element of the array as the pivot element.
 * - Compute the number of comparisons, always using the final element of the given array as the pivot element.
 *   Recall from the lectures that, just before the main Partition subroutine, you should exchange the pivot element (i.e., the last element) with the first element.
 * - Compute the number of comparisons, using the "median-of-three" pivot rule. 
 *   [This primary motivation behind this rule is to do a little bit of extra work to get much better performance on input arrays that are already sorted.] 
 *   In more detail, you should choose the pivot as follows. Consider the first, middle, and final elements of the given array. 
 *   (If the array has odd length it should be clear what the "middle" element is; for an array with even length 2k, use the k-th element as the "middle" element. 
 *   So for the array: 4 5 6 7, the "middle" element is the second one ---- 5 and not 6!) 
 *   Identify which of these three elements is the median (i.e., the one whose value is in between the other two), and use this as your pivot. 
 *   SUBTLE POINT: A careful analysis would keep track of the comparisons made in identifying the median of the three elements. 
 *   You should NOT do this. That is, as in the previous two problems, you should simply add m−1 to your running total of comparisons every time you recurse on a subarray with length m.
 * 
 * @author ddudnik
 *
 */
public class CountQuickSortComparisons {
	
	private static BigDecimal comparisonCounter = new BigDecimal(0);
	
	public static enum PivotSelection{
		FIRST,
		LAST,
		MEDIAN
	}

	public static void main(String[] args) throws IOException {
		String fName = args[0];
		if (fName == null || fName.isEmpty()){
			throw new IllegalArgumentException("File name should be specified as the only one argument!");
		}
		int[] input = initializeInputArray(fName);
		
		CountQuickSortComparisons.comparisonCounter = new BigDecimal(0);
		performQuickSort(input.clone(), PivotSelection.FIRST);
		System.out.println("First element selected as pivot, comparisons count: " + CountQuickSortComparisons.comparisonCounter.longValueExact());
		
		CountQuickSortComparisons.comparisonCounter = new BigDecimal(0);
		performQuickSort(input.clone(), PivotSelection.LAST);
		System.out.println("Last element selected as pivot, comparisons count: " + CountQuickSortComparisons.comparisonCounter.longValueExact());

		CountQuickSortComparisons.comparisonCounter = new BigDecimal(0);
		performQuickSort(input.clone(), PivotSelection.MEDIAN);
		System.out.println("Median of three element selected as pivot, comparisons count: " + CountQuickSortComparisons.comparisonCounter.longValueExact());
	}
	
	private static int[] performQuickSort(int[] input, PivotSelection pivotSelection){
		if (input.length <= 1){
			return input;
		}
		
		int pivotIndex = 0;
		
		switch (pivotSelection) {
		case FIRST:
			pivotIndex = 0;
			break;
		case LAST:
			pivotIndex = input.length - 1;
			break;
		case MEDIAN:
			int firstElement = input[0];
			int lastElement = input[input.length - 1];
			int middleElement = 0;
			int middleIndex = 0;
			if (input.length % 2 == 0){
				middleIndex = input.length / 2;
			}else{
				middleIndex = (input.length + 1) / 2; 
			}
			middleIndex--;
			middleElement = input[middleIndex];
			
			int[] three = new int[]{firstElement, middleElement, lastElement};
			Arrays.sort(three);
			int pivot = three[1];
			
			if (pivot == firstElement){
				pivotIndex = 0;
			}else if (pivot == lastElement){
				pivotIndex = input.length - 1;
			}else{
				pivotIndex = middleIndex;
			}
			
			break;
		default:
			throw new RuntimeException("Wrong pivot selection!");
		}

		// counting comparisons
		comparisonCounter = comparisonCounter.add(new BigDecimal(input.length - 1));

		int pivotElement = input[pivotIndex];
		partition(input, pivotIndex);
		
		Collection<Integer> first = new ArrayList<Integer>();
		Collection<Integer> second = new ArrayList<Integer>();
		for (int element : input){
			if (element < pivotElement){
				first.add(element);
			}else if (element > pivotElement){
				second.add(element);
			}
		}
		
		int[] firstArray = collectionToArray(first);
		int[] secondArray = collectionToArray(second);
		
		int[] firstSorted = performQuickSort(firstArray, pivotSelection);
		int[] secondSorted = performQuickSort(secondArray, pivotSelection);
		
		int[] firstSortedNew = Arrays.copyOf(firstSorted, firstSorted.length + 1);
		firstSortedNew[firstSortedNew.length - 1] = pivotElement;
		
		return concat(firstSortedNew, secondSorted);
	}
	
	private static void partition(int[] input, int pivotIndex){

		int pivot = input[pivotIndex];
		if (pivotIndex > 0){
			swap(input, 0, pivotIndex);
		}
		
		int i = 1;
		for (int j = 1; j < input.length; j++){
			if (input[j] < pivot){
				swap(input, j, i);
				i++;
			}
		}
		
		swap(input, 0, i -1);
	}
	
	private static int[] collectionToArray(Collection<Integer> collection){
		int[] result = new int[collection.size()];
		int i = 0;
		for (Integer element : collection){
			result[i] = element.intValue();
			i++;
		}
		return result;
	}
	
	private static int[] concat(int[] A, int[] B) {
		int[] C = new int[A.length + B.length];
		System.arraycopy(A, 0, C, 0, A.length);
		System.arraycopy(B, 0, C, A.length, B.length);
		return C;
	}
	
	private static void swap(int[] input, int i, int j){
		int temp = input[i];
		input[i] = input[j];
		input[j] = temp;
	}
	
	private static int[] initializeInputArray(String fileName) throws NumberFormatException, IOException{
		FileInputStream fis = new FileInputStream(fileName);
		DataInputStream in = new DataInputStream(fis);
		BufferedReader reader = new BufferedReader(new InputStreamReader(in));
		
		int[] input = new int[10000]; 
		String strLine;
		int i = 0;
		while ((strLine = reader.readLine()) != null)   {
			input[i] = Integer.parseInt(strLine);
			i++;
		}
		in.close();
		
		return input;
	}
}

File: workspace/algo-class-assignments/ruby/main/week2/count_quicksort_comparisons.rb
# The file contains all of the integers between 1 and 10,000 (inclusive) in unsorted order (with no integer repeated). 
# The integer in the i-th row of the file gives you the i-th entry of an input array.
# Your task is to compute the total number of comparisons used to sort the given input file by QuickSort. 
# As you know, the number of comparisons depends on which elements are chosen as pivots, so we'll ask you to explore three different pivoting rules.
# You should not count comparisons one-by-one. Rather, when there is a recursive call on a subarray of length m, 
# you should simply add m−1 to your running total of comparisons. (This is because the pivot element will be compared to each of the other m−1 elements in the subarray in this recursive call.)
# 
# WARNING: The Partition subroutine can be implemented in several different ways, and different implementations can give you differing numbers of comparisons. 
# For this problem, you should implement the Partition subroutine as it is described in the video lectures (otherwise you might get the wrong answer).
# 
# DIRECTIONS FOR PROBLEMS:
# - For the first part of the programming assignment, you should always use the first element of the array as the pivot element.
# - Compute the number of comparisons, always using the final element of the given array as the pivot element.
#   Recall from the lectures that, just before the main Partition subroutine, you should exchange the pivot element (i.e., the last element) with the first element.
# - Compute the number of comparisons, using the "median-of-three" pivot rule. 
#   [This primary motivation behind this rule is to do a little bit of extra work to get much better performance on input arrays that are already sorted.] 
#   In more detail, you should choose the pivot as follows. Consider the first, middle, and final elements of the given array. 
#   (If the array has odd length it should be clear what the "middle" element is; for an array with even length 2k, use the k-th element as the "middle" element. 
#   So for the array: 4 5 6 7, the "middle" element is the second one ---- 5 and not 6!) 
#   Identify which of these three elements is the median (i.e., the one whose value is in between the other two), and use this as your pivot. 
#   SUBTLE POINT: A careful analysis would keep track of the comparisons made in identifying the median of the three elements. 
#   You should NOT do this. That is, as in the previous two problems, you should simply add m−1 to your running total of comparisons every time you recurse on a subarray with length m.

$comparison_counter = 0

def initialize_input_array(filename)
  input = []
  File.open(filename, 'r') do |file|
    while line = file.gets
      input.push line.to_i 10
    end
  end
  input
end

class PivotSelection
  FIRST=1
  LAST=2
  MEDIAN=3
end

class Array

  def swap!(a, b)
         self[a], self[b] = self[b], self[a]
    self
  end
    
  def quicksort_partition!(pivot_index)
    raise IndexError, "#{pivot_index} >= size (#{self.size}) !!!" if pivot_index >= self.size
    
    pivot = self[pivot_index]
    if pivot_index > 0
      self.swap!(0, pivot_index)
    end
    
    split_index = 1;
    self.each_with_index do |item, index|
      if index == 0
        next
      end
      if item < pivot
        self.swap!(index, split_index)
        split_index += 1
      end
    end
    
    self.swap!(0, split_index - 1)
    
    self
 end
end

def perform_quicksort(input_array, pivot_selection)
    return input_array if input_array.size <= 1
    
    if PivotSelection::FIRST == pivot_selection
      pivot_index = 0
    elsif PivotSelection::LAST == pivot_selection
      pivot_index = input_array.size - 1
    elsif PivotSelection::MEDIAN == pivot_selection
      first = input_array[0]
      last = input_array[input_array.size - 1]
      middle = 0
      middle_index = 0
      if input_array.size.even?
        middle_index = input_array.size/2
      else
        middle_index = (input_array.size + 1)/2
      end
      middle_index -= 1
      middle = input_array[middle_index]
      
      pivot = [first, middle, last].sort[1]
      if pivot == first
        pivot_index = 0
      elsif pivot == last
        pivot_index = input_array.size - 1
      else
        pivot_index = middle_index
      end
      
    else
      raise "Wrong pivot selection!"
    end
    
    pivot_element = input_array[pivot_index]
    
    # counting comparisons
    $comparison_counter += input_array.size - 1

    begin
      input_array.quicksort_partition! pivot_index
    rescue Exception => e
      puts "Wrong pivot index!"
      puts e.message
      puts e.backtrace.inspect
      return input_array  
    end
    
    first, second = input_array.partition do |item|
      item < pivot_element
    end
    second.shift
    
    sorted_first = perform_quicksort(first, pivot_selection)
    
    sorted_second = perform_quicksort(second, pivot_selection)
    
    sorted_first.push(pivot_element)
    sorted_first + sorted_second
end

if (ARGV.size > 0)
  input_array = initialize_input_array ARGV[0]

  $comparison_counter = 0
  perform_quicksort(Marshal.load(Marshal.dump(input_array)), PivotSelection::FIRST)
  puts "First element selected as pivot, comparisons count: #{$comparison_counter}"
  
  $comparison_counter = 0
  perform_quicksort(Marshal.load(Marshal.dump(input_array)), PivotSelection::LAST)
  puts "Last element selected as pivot, comparisons count: #{$comparison_counter}"

  $comparison_counter = 0
  perform_quicksort(Marshal.load(Marshal.dump(input_array)), PivotSelection::MEDIAN)
  puts "Median of three selected as pivot, comparisons count: #{$comparison_counter}"
  
else
  puts "File name should be specified as the only one argument!"
end